Online Food Ordering Example (#289)
serverlessworkflow/specification: examples/README.md

</td>
</tr>
</table>

### Online Food Ordering

#### Description

> In this example we want to create an online food ordering workflow. Below image outlines the workflow 
> structure and the available services:

Reviewer: ```suggestion
In this example, we want to create an online food ordering workflow. The below image outlines the workflow 
structure and the available services:
```

<p align="center">
<img src="../media/examples/example-foodorder-outline.png" height="450px" alt="Online Food Ordering Structure"/>
</p>

> Our workflow starts off with the "Place Order" [Subflow](../specification.md#SubFlow-State) which is responsible
> to send the received order to the requested restaurant and the estimated order ETA. 
> We then wait for the ETA time when our workflow should go into the "Deliver Order" SubFlow which is responsible
> for dispatching a Courier and send her/him off to pick up the order. Once the order is picked up the Courier 
> needs to deliver the order to the customer. 
> After the order has been delievered to the customer, our workflow needs to charge the customer for the order. 
> 
> Our workflow needs to communicate with three services during its execution, namely the Order, Delivery, and 
> the Payment services. 
> 
> For the sake of the example we assume that our workflow can communicate to the Order and Delivery services via REST
> and the Payment service via gRPC. 
> 
> Let's start off by defining an example CloudEvent which triggers an instance of our workflow.
> This event can be sent by a web UI for example, or be pushed onto a Kafka/MQTT topic to start 
> our order workflow.

Reviewer: ```suggestion
Our workflow starts with the "Place Order" [Subflow](../specification.md#SubFlow-State), which is responsible
to send the received order to the requested restaurant and the estimated order ETA. 
We then wait for the ETA time when our workflow should go into the "Deliver Order" SubFlow, responsible
for dispatching a Courier and sending her/him off to pick up the order. Once the order is picked up, the Courier needs to deliver the order to the customer. 
After the order has been delivered to the customer, our workflow needs to charge the customer. 

Our workflow needs to communicate with three services during its execution, namely the Order, Delivery, and 
the Payment services. 

For the sake of the example, we assume that our workflow can communicate to the Order and Delivery services via REST and the Payment service via gRPC. 

Let's start by defining an example CloudEvent which triggers an instance of our workflow.
This event can be sent by a web UI, for example, or be pushed onto a Kafka/MQTT topic to start our order workflow.
```

```json
{
   "specversion": "1.0",
   "type": "org.orders",
   "source": "/orders/",
   "subject": "Food Order",
   "id": "A234-1234-1234",
   "time": "2021-03-05T17:31:00Z",
   "orderid": "ORDER-12345",
   "data": {
      "id": "ORDER-12345",
      "customerId": "CUSTOMER-12345",
      "status": [],
      "order": {
         "restaurantId": "RESTAURANT-54321",
         "items": [
            {
               "itemId": "ITEM-8765",
               "amount": 1,
               "addons": ""
            }
         ]
      },
      "delivery":{
         "address": "1234 MyStreet, MyCountry",
         "type": "contactless",
         "requestedTime": "ASAP",
         "location": "Front door",
         "instructions": ""
      }
   }
}
```

> Note the "orderid" CloudEvent context attribute which contains the unique ID of the order specified 
> in this event. [Event correlation](../specification.md#Correlation-Definition) is done against CE context attributes, and as such to be able
> to correlate multiple order events to the same order id, it needs to be part of the CE context attributes, and 
> not its data (payload).  

Reviewer: ```suggestion
Note the `orderid` CloudEvent context attribute, which contains the unique ID of the order specified in this event. [Event correlation](../specification.md#Correlation-Definition) is done against CE context attributes, and as such, to be able
to correlate multiple order events to the same order id, it needs to be part of the CE context attributes, and 
not its data (payload).  
```

> Now let's start defining our workflow. For the sake of this example let's define our function and event definitions
> as separate YAML files (and then reference them inside our workflow definition). This is useful in cases
> when you want to reuse them between multiple workflow definitions.

Reviewer: ```suggestion
Now let's start defining our workflow. For the sake of this example, let's define our function and event definitions
as separate YAML files (and then reference them inside our workflow definition). This is useful in cases
when you want to reuse them between multiple workflow definitions.
```

        results: "${ .results.status }"
        toStateData: "${ .status }"
    - functionRef:
        refName: Get Order ETA Function
        arguments:
          customer: "${ .customerId }"
          restaurantid: "${ .order.restaurantId }"
          delivery: " ${ .delivery }"
      actionDataFilter:
        results: "${ .results.status }"
        toStateData: "${ .status }"
  end: true
```

#### Deliver Order Sub-Workflow

``` yaml
id: deliverorderworkflow
name: Deliver Order Workflow
version: '1.0'
start: Dispatch Courier
states:
- name: Dispatch Courier
  type: operation
  actions:
  - functionRef: Dispatch Courrier Function
  transition: Wait for Order Pickup
- name: Wait for Order Pickup
  type: event
  onEvents:
  - eventRefs:
    - Order Picked Up Event
    eventDataFilter:
      data: "${ .data.status }"
      toStateData: "${ .status }"
    actions:
    - functionRef: Deliver Order Function
  transition: Wait for Delivery Confirmation
- name: Wait for Delivery Confirmation
  type: event
  onEvents:
  - eventRefs:
    - Order Delievered Event
    eventDataFilter:
      data: "${ .data.status }"
      toStateData: "${ .status }"
  end: true
```

#### Workflow Results

> For the example order event the workflow output for a successful completion would look like for example:

Reviewer: ```suggestion
For the example order event, the workflow output for a successful completion would look like for example:
```
